contributors/design-proposals/live-and-inplace-vertical-scaling.md

```diff
@@ -0,0 +1,325 @@
+
+# Motivation
+
+Resources needed by containers can change over time for a variety of reasons - moving from live-test mode to production usage, change in user load or dataset sizes each of which again might come about for a variety of reasons.
+`Statefulset` supports the capability to change Request and Limit values specified for a container through supported pod spec update methods.
+However, they currently require the pods be restarted to run with the new resource sizes.
+
+There are a few reasons why restarting may not be desirable particularly for stateful services.
+
+1. Moving or copying existing state information or re-sharding employing horizontal scaling may be expensive or even infeasible depending on the service or underlying application implementation.
+
+2. Services with large amount of state associated with specific instances may encounter temporary loss in access to the specific state information while its corresponding container is being restarted. Or they might encounter non-trivial delays in getting back to expected performance.
+
+3. State may be maintained on node-local storage/devices and restart might place the container/pod on a different node (unless suitably constrained). Might create more fragile scaling option for service and/or additional burden (to constrain placement) on user.
+
+4. Where the restart is not needed, unnecessary load may be created on Kubernetes components that are responsible for pod deletion and creation potentially impacting Kubernetes scalability.
+Resizing running containers of a pod in place could avoid that additional load.
+
+Conversely, resizing without restart may not be the right option for all applications and the right option may also be resource dependent.
+For example, an application might (or have the means to) assess resources available to it only at its initialization time and be unable to adjust its usage if those are dynamically changed.
+Or it might be able to adjust to changes in certain resources (most can for CPU) but not for others (some application have particular difficulty freeing memory they have taken up).
+So we need the means to express and implement the best approach for resizing for each workload and resource.
+
+# Use Cases
+
+* After deploying a service/application through `StatefulSet`, I want to resize, by giving a updated pod spec with new resource sizes, the resources (for now, CPU and memory) allocated to each pod of the statefulset without restarting the pods.
+
+* A (vertical) pod auto-scaler (c.f. [a related issue](https://github.com/kubernetes/features/issues/21)) can utilize this feature, while currently it does resize a pod with restart only.
+
+# Objectives
+
+1. Enable live and in-place resource resizing on a pod.
+2. Add support for live and in-place resource resizing in `StatefulSet` controller.
+
+# API and Usage
+
+To express the policy for resizing in a pod spec, we introduce resource attribute `resizePolicy` with the following choices for value:
```

Comment:
Maybe I missed the discussion, but do we really need this to be so configurable?  Or can we make the default "say nothing" be the best?  E.g. If the user says nothing, always try to resize in place.  If that can't be achieved, which can't always be known statically, then do a pod restart.

If the user NEEDS a restart, they can specify that, but make it the abnormal case.

?

Reply:
Depending on the nature of workloads and resources, their feasible policy is different so the policy needs to be configurable by users. For example, we can't resize the size of memory allocated to JVM dynamically, so in this case, the policy needs to be restartOnly. Also, some workload doesn't tolerate restarts, so in such case, the policy is to be configured as LiveResizeOnly.